### ISSUE TITLE
What is the shape of the graph network input layer that represents a single crystal structure?

### ISSUE THREAD

**Issue by sgbaird** [txie-93/cdvae#13]
I believe CGCNN is a 3D matrix that looks more like a cube than a high-aspect ratio box, though I don't remember the specific default shape off the top of my head. Assuming this is a valid question to be asking, what is the input shape that represents a single crystal structure in CDVAE? E.g. $100\times100\times50$.

**Comment by kyonofx** [comment]
Hi,

Each crystal structure is represented as a graph, so there is not a fixed shape for an input crystal.

How we construct graphs: https://github.com/txie-93/cdvae/blob/main/cdvae/common/data_utils.py#L108

We use PyG's data objects: https://github.com/txie-93/cdvae/blob/main/cdvae/pl_data/dataset.py#L57

Hope this helps.

**Comment by sgbaird** [comment]
@kyonofx,

Based on the links you shared, I came across https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html?highlight=padding#advanced-mini-batching which mentions how concatenation occurs without padding.

> The creation of mini-batching is crucial for letting the training of a deep learning model scale to huge amounts of data. Instead of processing examples one-by-one, a mini-batch groups a set of examples into a unified representation where it can efficiently be processed in parallel. In the image or language domain, this procedure is typically achieved by rescaling or padding each example into a set to equally-sized shapes, and examples are then grouped in an additional dimension. The length of this dimension is then equal to the number of examples grouped in a mini-batch and is typically referred to as the batch_size.
>
> Since graphs are one of the most general data structures that can hold any number of nodes or edges, the two approaches described above are either not feasible or may result in a lot of unnecessary memory consumption. In PyG, we opt for another approach to achieve parallelization across a number of examples. Here, adjacency matrices are stacked in a diagonal fashion (creating a giant graph that holds multiple isolated subgraphs), and node and target features are simply concatenated in the node dimension, i.e.
>
> ![image](https://user-images.githubusercontent.com/45469701/173199157-3a33899f-d8ed-418b-9d55-ca572ae81733.png)
>
> This procedure has some crucial advantages over other batching procedures:
> 
> 1. GNN operators that rely on a message passing scheme do not need to be modified since messages still cannot be exchanged between two nodes that belong to different graphs.
> 
> 1. There is no computational or memory overhead. For example, this batching procedure works completely without any padding of node or edge features. Note that there is no additional memory overhead for adjacency matrices since they are saved in a sparse fashion holding only non-zero entries, i.e., the edges.

https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.Data helps out with understanding some of the shapes of things on a per-crystal basis.

Thanks!

**Comment by txie-93** [comment]
The key about batching is here: https://github.com/txie-93/cdvae/blob/f857f598d6f6cca5dc1ea0582d228f12dcc2c2ea/cdvae/pl_data/dataset.py#L57-L68

We use the trick that pytorch geometric by default concatenates the 0th dimention during batching.

For `frac_coords`, `atom_types`, `edge_index`, `to_jimages`, they are node/edge level features. The batching considers them in a big graph where each crystal is a disconnected subgraph.

For `lengths`, `angles`, `y`, they are graph level properties so we unsqueeze a 0th dimension of size 1. 

Does this answer your question?

**Comment by sgbaird** [comment]
@txie-93 I think that makes sense and yes, answers the question. Thank you!
